### DISCUSSION TITLE
Getting the sender of an event

### DISCUSSION THREAD

**Discussion by programmingkidx** [progrium/darwinkit#162] [General]
In Objective-C the action method to a control like a button has a sender argument that lets the method know who sent the event. Is something like this possible in MacDriver? I made a program that display buttons in a grid and I want to print out the name of the button that generated an event. Here is the program: 

```
package main

import "github.com/progrium/macdriver/cocoa"
import "github.com/progrium/macdriver/objc"
import "github.com/progrium/macdriver/core"
import "fmt"

func main() {

	myclass := objc.NewClass("MyClass", "NSObject")
	myclass.AddMethod("doButton:", doButton)
	objc.RegisterClass(myclass)
	myobj := objc.Get("MyClass").Alloc()
	
	
	app := cocoa.NSApp_WithDidLaunch(func(n objc.Object) {
			win := cocoa.NSWindow_New()
			win.SetTitle("NSButton")
			winRect := core.NSMakeRect(20, 20, 700, 300)
			win.SetFrameDisplay(winRect, true)
			win.MakeKeyAndOrderFront(nil)

			/* NSButton code */
			buttonWidth := 70.0
			buttonHeight := 30.0
			for y := 0; y < 10; y++ {
				for x := 0; x < 10; x++ {
					rect := core.NSMakeRect(float64(x) * buttonWidth, float64(y) * buttonHeight, buttonWidth, buttonHeight)
					button := cocoa.NSButton_Alloc()
					button.InitWithFrame(rect)
					newTitle := fmt.Sprintf("%dx%d", x, y)
					button.SetTitle(newTitle)
					button.SetTarget(myobj)
					button.SetAction(objc.Sel("doButton:"))
					win.ContentView().AddSubview(button)
				}
			}
		})

		app.SetActivationPolicy(cocoa.NSApplicationActivationPolicyRegular)
		app.ActivateIgnoringOtherApps(true)
		app.Run()
}

func doButton(button objc.Object) {
	fmt.Println("Button pressed:", button)
}

```

The window this program displays looks like this:
<img width="812" alt="grid" src="https://github.com/progrium/macdriver/assets/19289541/d5d6ddb4-392a-4125-aa93-be0d9506401a">

Can the doButton() function's argument somehow tell who sent the event? Currently the argument is set to a MyClass class instance.

**Comment by progrium** [comment]
When you use AddMethod, the first argument of the callback will be the instance of the class you added this method to. The second (I believe) will be the selector. The rest will be the arguments to the method, which I believe is what you're looking for. Try adding two more arguments and using the third.
